File: OJ/lcof38/lcof38.cpp
/*
lcof38


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <random>
#include <functional>
#include <utility>
#include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    vector<string> permutation(string s) {
        auto n = s.size();
        vector<string> ans;
        vector<bool> used(n);
        string path;

        sort(s.begin(), s.end());

        function<void()> dfs = [&]() {
            if (n == path.size()) {
                ans.push_back(path); return;
            }

            for (size_t i = 0; i < n; i++)
            {
                if (used[i]) continue;
                 // only visit the first repeated nodes
                if (i > 0 and s[i] == s[i-1] and not used[i-1]) continue;

                used[i] = true;
                path.push_back(s[i]);
                dfs();
                path.pop_back();
                used[i] = false;
            }
        };

        dfs();
        return ans;
    }

};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.permutation("abbcc") << endl;
    return 0;
}

File: OJ/lcof38/lcof38.py
# lcof38
# %%
from typing import List
from pprint import pprint

def init(dims, filled_value = 0):
    if not dims: return filled_value
    return [init(dims[1:], filled_value) for _ in range(dims[0])]


class Solution:
    def permutation(self, s: str) -> List[str]:
        N = len(s)
        s = sorted(s)
        path = []
        used = [False]  * N 
        ans = []
        def dfs():
            if len(path) == N:
                ans.append(''.join(path))
                return 
            
            for i in range(N):
                if used[i]: continue
                if i > 0 and s[i] == s[i-1] and not used[i-1]: continue

                used[i] = True
                path.append(s[i])
                dfs()
                path.pop()
                used[i] = False
        
        dfs()
        return ans


pprint(Solution().permutation("caba"))


